src/QMCWaveFunctions/TWFPrototype.cpp

```diff
@@ -0,0 +1,316 @@
+//////////////////////////////////////////////////////////////////////////////////////
+// This file is distributed under the University of Illinois/NCSA Open Source License.
+// See LICENSE file in top directory for details.
+//
+// Copyright (c) 2021 QMCPACK developers.
+//
+// File developed by:   Raymond Clay III, rclay@sandia.gov, Sandia National Laboratories
+//
+// File created by:   Raymond Clay III, rclay@sandia.gov, Sandia National Laboratories
+//////////////////////////////////////////////////////////////////////////////////////
+
+#include "QMCWaveFunctions/TWFPrototype.h"
+#include "Numerics/DeterminantOperators.h"
+#include <iostream>
+namespace qmcplusplus
+{
+TWFPrototype::TWFPrototype() : initialized(false) { std::cout << "TWFPrototype Constructed\n"; }
+
+TWFPrototype::IndexType TWFPrototype::get_group_index(const IndexType gid)
+{
+  IndexType ndets = groups.size();
+  for (IndexType i = 0; i < ndets; i++)
+    if (gid == groups[i])
+      return i;
+
+  return IndexType(-1);
+}
+void TWFPrototype::add_determinant(const ParticleSet& P, const IndexType gid, SPOSet* spo)
+{
+  if (std::find(groups.begin(), groups.end(), gid) == groups.end())
+  {
+    groups.push_back(gid);
+    spos.push_back(spo);
+    IndexType first = P.first(gid);
+    IndexType last  = P.last(gid);
+    IndexType norbs = spo->getOrbitalSetSize();
+    num_orbs.push_back(norbs);
+    num_ptcls.push_back(last - first);
+  }
+  initialized = true;
+}
+
+TWFPrototype::IndexType TWFPrototype::get_det_id(const IndexType species_id)
+{
+  IndexType detIndex = -1;
+  for (IndexType i = 0; i < groups.size(); i++)
+  {
+    if (species_id == groups[i])
+      detIndex = i;
+  }
+  assert(detIndex != -1);
+
+  return detIndex;
+}
+
+void TWFPrototype::get_M(const ParticleSet& P, std::vector<ValueMatrix_t>& mvec)
+{
+  IndexType ndets  = spos.size();
+  IndexType norbs  = 0;
+  IndexType nptcls = 0;
+  IndexType gid    = 0;
+  IndexType first  = 0;
+  IndexType last   = 0;
+  for (IndexType i = 0; i < ndets; i++)
+  {
+    gid     = groups[i];
+    first   = P.first(i);
+    last    = P.last(i);
+    nptcls  = last - first;
```

Comment:
Also available as object state. although I agree you shouldn't use the state.

Reply:
I think Peter means moving the definition inside the for loop. They are all scalars. So performance wise, it is better inside the loop. The code also gets more concise.

Reply:
Done.